src/equations/bbm_1d.jl

```diff
@@ -0,0 +1,241 @@
+
+See section 4.1.3 in (there is an error in paper: it should be `sech^2` instead of `cosh`):
+- Hendrik Ranocha, Dimitrios Mitsotakis and David I. Ketcheson (2020)
+  A Broad Class of Conservative Numerical Methods for Dispersive Wave Equations
+  [DOI: 10.4208/cicp.OA-2020-0119](https://doi.org/10.4208/cicp.OA-2020-0119)
+"""
+function initial_condition_convergence_test(x, t, equations::BBMEquation1D, mesh)
+    g = gravity_constant(equations)
+    D = equations.D
+    alpha = sqrt(g * D)
+    beta = 3 / 2 * sqrt(g / D)
+    gamma = 1 / 6 * D^2
+    c = 1.2 * alpha
+    A = 3 * (c - alpha) / beta
+    K = 0.5 * sqrt(1 / gamma * (1 - alpha / c))
+    x_t = mod(x - c * t - xmin(mesh), xmax(mesh) - xmin(mesh)) + xmin(mesh)
+    eta = A * sech(K * x_t)^2
+    return SVector(eta)
+end
+
+"""
+    initial_condition_manufactured(x, t, equations::BBMEquation1D, mesh)
+
+A smooth manufactured solution in combination with [`source_terms_manufactured`](@ref).
+"""
+function initial_condition_manufactured(x, t,
+                                        equations::BBMEquation1D,
+                                        mesh)
+    eta = exp(t / 2) * sinpi(2 * (x - t / 2))
+    return SVector(eta)
+end
+
+"""
+    source_terms_manufactured(q, x, t, equations::BBMEquation1D, mesh)
+
+A smooth manufactured solution in combination with [`initial_condition_manufactured`](@ref).
+"""
+function source_terms_manufactured(q, x, t, equations::BBMEquation1D)
+    g = gravity_constant(equations)
+    D = still_waterdepth(q, equations)
+    a1 = sqrt(g * D)
+    a2 = sqrt(g / D)
+    a3 = cospi(t - 2 * x)
+    a4 = sinpi(t - 2 * x)
+    a5 = sinpi(2 * t - 4 * x)
+    a6 = exp(t / 2)
+    dq1 = -pi^2 * D^2 * (a4 + 2 * pi * a3) * a6 / 3 - 3 * pi * a2 * exp(t) * a5 / 2 +
+          2 * pi * a1 * a6 * a3 - a6 * a4 / 2 - pi * a6 * a3
+
+    return SVector(dq1)
+end
+
+function create_cache(mesh, equations::BBMEquation1D,
+                      solver, initial_condition,
+                      ::BoundaryConditionPeriodic,
+                      RealT, uEltype)
+    D = equations.D
+    invImD2 = lu(I - 1 / 6 * D^2 * sparse(solver.D2))
+    eta2 = zeros(RealT, nnodes(mesh))
+    eta2_x = zero(eta2)
+    eta_x = zero(eta2)
+    etaeta_x = zero(eta2)
+    eta_xx = zero(eta2)
+    g = gravity_constant(equations)
+    c_0 = sqrt(g * D)
+    c_1 = sqrt(g / D)
+    cache = (; invImD2, eta2, eta2_x, eta_x, etaeta_x, eta_xx, c_0, c_1,
+             solver.D1, solver.D2)
+    if solver.D1 isa PeriodicUpwindOperators
+        eta_x_upwind = zero(eta2)
+        cache = (; cache..., eta_x_upwind)
+    end
+    return cache
+end
+
+# Discretization that conserves the mass for eta and the modified energy for periodic boundary conditions, see
+# - Hendrik Ranocha, Dimitrios Mitsotakis and David I. Ketcheson (2020)
+#   A Broad Class of Conservative Numerical Methods for Dispersive Wave Equations
+#   [DOI: 10.4208/cicp.OA-2020-0119](https://doi.org/10.4208/cicp.OA-2020-0119)
```

Comment:
This needs to be updated based since we can have a split form or not.

Reply:
[05d3fbd](https://github.com/JoshuaLampert/DispersiveShallowWater.jl/pull/150/commits/05d3fbd740d070a5229a0f9d457bfb341353264d)